File: api/lua/pinnacle/signal.lua
-- This Source Code Form is subject to the terms of the Mozilla Public
-- License, v. 2.0. If a copy of the MPL was not distributed with this
-- file, You can obtain one at https://mozilla.org/MPL/2.0/.

local log = require("pinnacle.log")
local client = require("pinnacle.grpc.client").client
local signal_service = require("pinnacle.grpc.defs").pinnacle.signal.v1.SignalService

local stream_control = require("pinnacle.grpc.defs").pinnacle.signal.v1.StreamControl

local signals = {
    OutputConnect = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(output: pinnacle.output.OutputHandle) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    OutputDisconnect = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(output: pinnacle.output.OutputHandle) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    OutputSetup = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(output: pinnacle.output.OutputHandle) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    OutputResize = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(output: pinnacle.output.OutputHandle, logical_width: integer, logical_height: integer) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    OutputMove = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(output: pinnacle.output.OutputHandle, x: integer, y: integer) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    OutputPointerEnter = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(output: pinnacle.output.OutputHandle) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    OutputPointerLeave = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(output: pinnacle.output.OutputHandle) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    OutputFocused = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(output: pinnacle.output.OutputHandle) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    WindowPointerEnter = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(window: pinnacle.window.WindowHandle) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    WindowPointerLeave = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(window: pinnacle.window.WindowHandle) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    WindowFocused = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(window: pinnacle.window.WindowHandle) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    WindowTitleChanged = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(window: pinnacle.window.WindowHandle, title: string) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    TagActive = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(tag: pinnacle.tag.TagHandle, active: boolean) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
    InputDeviceAdded = {
        ---@type grpc_client.h2.Stream?
        sender = nil,
        ---@type { callback_id: integer, callback: fun(device: pinnacle.input.libinput.DeviceHandle) }[]
        callbacks = {},
        ---@type fun(response: table)
        on_response = nil,
    },
}

---Call a signal callback in protected mode
---
---If the signal fails, the error is logged.
---The handle parameter is expected to be the first parameter of the callback, and to be convertible
---to a string via tostring(). If it's not the case, setting this parameter to nil will not pass it
---to the callback, and will not uses in in the error string generation.
---
---@generic T
---@param signal_name string
---@param callback function
---@param handle T|nil Either a Handle object that'll be passed as callback first arg, or nil
local function protected_callback(signal_name, callback, handle, ...)
    local success, err

    if handle then
        success, err = pcall(callback, handle, ...)
    else
        success, err = pcall(callback, ...)
    end

    if not success then
        local errstr = "While handling '" .. signal_name .. "'"
        if handle then
            errstr = errstr .. " for " .. tostring(handle)
        end
        errstr = errstr .. ": " .. tostring(err)

        log.error(errstr)
    end
end

-- NOTE: We need to copy callbacks into a new table because callbacks are able to disconnect signals
-- while we iterate through them, which is a no-no

signals.OutputConnect.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local handle = require("pinnacle.output").handle.new(response.output_name)
    local callbacks = require("pinnacle.util").deep_copy(signals.OutputConnect.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("OutputConnect", callback.callback, handle)
    end
end

signals.OutputDisconnect.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local handle = require("pinnacle.output").handle.new(response.output_name)
    local callbacks = require("pinnacle.util").deep_copy(signals.OutputDisconnect.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("OutputDisconnect", callback.callback, handle)
    end
end

signals.OutputSetup.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local handle = require("pinnacle.output").handle.new(response.output_name)
    local callbacks = require("pinnacle.util").deep_copy(signals.OutputSetup.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("OutputSetup", callback.callback, handle)
    end
end

signals.OutputResize.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local handle = require("pinnacle.output").handle.new(response.output_name)
    local callbacks = require("pinnacle.util").deep_copy(signals.OutputResize.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback(
            "OutputResize",
            callback.callback,
            handle,
            response.logical_width,
            response.logical_height
        )
    end
end

signals.OutputMove.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local handle = require("pinnacle.output").handle.new(response.output_name)
    local callbacks = require("pinnacle.util").deep_copy(signals.OutputMove.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("OutputMove", callback.callback, handle, response.x, response.y)
    end
end

signals.OutputPointerEnter.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local handle = require("pinnacle.output").handle.new(response.output_name)
    local callbacks = require("pinnacle.util").deep_copy(signals.OutputPointerEnter.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("OutputPointerEnter", callback.callback, handle)
    end
end

signals.OutputPointerLeave.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local handle = require("pinnacle.output").handle.new(response.output_name)
    local callbacks = require("pinnacle.util").deep_copy(signals.OutputPointerLeave.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("OutputPointerLeave", callback.callback, handle)
    end
end

signals.OutputFocused.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local handle = require("pinnacle.output").handle.new(response.output_name)
    local callbacks = require("pinnacle.util").deep_copy(signals.OutputFocused.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("OutputFocused", callback.callback, handle)
    end
end

signals.WindowPointerEnter.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local window_handle = require("pinnacle.window").handle.new(response.window_id)
    local callbacks = require("pinnacle.util").deep_copy(signals.WindowPointerEnter.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("WindowPointerEnter", callback.callback, window_handle)
    end
end

signals.WindowPointerLeave.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local window_handle = require("pinnacle.window").handle.new(response.window_id)
    local callbacks = require("pinnacle.util").deep_copy(signals.WindowPointerLeave.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("WindowPointerLeave", callback.callback, window_handle)
    end
end

signals.WindowFocused.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local window_handle = require("pinnacle.window").handle.new(response.window_id)
    local callbacks = require("pinnacle.util").deep_copy(signals.WindowFocused.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("WindowFocused", callback.callback, window_handle)
    end
end

signals.WindowTitleChanged.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local window_handle = require("pinnacle.window").handle.new(response.window_id)
    local callbacks = require("pinnacle.util").deep_copy(signals.WindowTitleChanged.callbacks)
    local title = response.title or ""

    for _, callback in ipairs(callbacks) do
        protected_callback("WindowTitleChanged", callback.callback, window_handle, title)
    end
end

signals.TagActive.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local tag_handle = require("pinnacle.tag").handle.new(response.tag_id)
    local callbacks = require("pinnacle.util").deep_copy(signals.TagActive.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("TagActive", callback.callback, tag_handle, response.active)
    end
end

signals.InputDeviceAdded.on_response = function(response)
    ---@diagnostic disable-next-line: invisible
    local device_handle = require("pinnacle.input.libinput").new_device(response.device_sysname)
    local callbacks = require("pinnacle.util").deep_copy(signals.InputDeviceAdded.callbacks)

    for _, callback in ipairs(callbacks) do
        protected_callback("InputDeviceAdded", callback.callback, device_handle)
    end
end

-----------------------------------------------------------------------------

---@class pinnacle.signal.SignalHandleModule
---@lcat nodoc
local signal_handle = {}

---A handle to a connected signal that can be used to disconnect the provided callback.
---
---@class pinnacle.signal.SignalHandle
---@lcat nodoc
---@field private signal string
---@lcat nodoc
---@field private callback_id integer The ID for the callback you connected
local SignalHandle = {}

---@class pinnacle.signal.SignalHandlesModule
---@lcat nodoc
local signal_handles = {}

---A collection of `SignalHandle`s retreived through a `connect_signal` function.
---@class pinnacle.signal.SignalHandles
local SignalHandles = {}

---@class pinnacle.signal.Signal
---@field private handle pinnacle.signal.SignalHandleModule
---@field private handles pinnacle.signal.SignalHandlesModule
---@lcat nodoc
local signal = {}
signal.handle = signal_handle
signal.handles = signal_handles

---Disconnect the provided callback from this signal.
function SignalHandle:disconnect()
    local cb_index = nil
    for i, cb in ipairs(signals[self.signal].callbacks) do
        if cb.callback_id == self.callback_id then
            cb_index = i
            break
        end
    end

    if cb_index then
        table.remove(signals[self.signal].callbacks, cb_index)
    end

    if #signals[self.signal].callbacks == 0 then
        signal.disconnect(self.signal)
    end
end

---@param request string
---@param callback_id integer
---@return pinnacle.signal.SignalHandle
function signal_handle.new(request, callback_id)
    ---@type pinnacle.signal.SignalHandle
    local self = {
        signal = request,
        callback_id = callback_id,
    }
    setmetatable(self, { __index = SignalHandle })
    return self
end

---Disconnects the callbacks from all the signal connections that are stored in this handle collection.
---
---@param self table<string, pinnacle.signal.SignalHandle>
function SignalHandles:disconnect_all()
    for _, sig in pairs(self) do
        sig:disconnect()
    end
end

---@return pinnacle.signal.SignalHandles
function signal_handles.new()
    ---@type pinnacle.signal.SignalHandles
    local self = {}
    setmetatable(self, { __index = SignalHandles })
    return self
end

local callback_ids = 0

---@param request string
---@param callback function
---@return pinnacle.signal.SignalHandle
---@lcat nodoc
function signal.add_callback(request, callback)
    if #signals[request].callbacks == 0 then
        signal.connect(request, signals[request].on_response)
    end

    local next_id = callback_ids
    callback_ids = callback_ids + 1

    table.insert(signals[request].callbacks, {
        callback_id = next_id,
        callback = callback,
    })

    return signal_handle.new(request, next_id)
end

---@param request string
---@param callback fun(response: table)
---@lcat nodoc
function signal.connect(request, callback)
    local stream = client:bidirectional_streaming_request(
        signal_service[request],
        function(response)
            callback(response)

            if signals[request].sender then
                local chunk = require("pinnacle.grpc.protobuf").encode(
                    "pinnacle.signal.v1." .. request .. "Request",
                    {
                        control = stream_control.STREAM_CONTROL_READY,
                    }
                )

                local success, err =
                    pcall(signals[request].sender.write_chunk, signals[request].sender, chunk)

                if not success then
                    print("error sending to stream:", err)
                    os.exit(1)
                end
            end
        end
    )

    signals[request].sender = stream

    local chunk =
        require("pinnacle.grpc.protobuf").encode("pinnacle.signal.v1." .. request .. "Request", {
            control = stream_control.STREAM_CONTROL_READY,
        })

    local success, err = pcall(signals[request].sender.write_chunk, signals[request].sender, chunk)
    if not success then
        print("error sending to stream:", err)
        os.exit(1)
    end
end

---This should only be called when call callbacks for the signal are removed
---@param request string
---@lcat nodoc
function signal.disconnect(request)
    if signals[request].sender then
        local chunk = require("pinnacle.grpc.protobuf").encode(
            "pinnacle.signal.v1." .. request .. "Request",
            {
                control = stream_control.STREAM_CONTROL_DISCONNECT,
            }
        )

        local success, err =
            pcall(signals[request].sender.write_chunk, signals[request].sender, chunk)
        if not success then
            print("error sending to stream:", err)
            os.exit(1)
        end

        signals[request].sender:shutdown()
        signals[request].sender = nil
    end
end

return signal

File: api/rust/src/signal.rs
//! Compositor signals.
//!
//! Your config can connect to various compositor signals that allow you to, for example, do
//! something when an output is connected or when the pointer enters a window.
//!
//! Some of the other modules have a `connect_signal` method that will allow you to pass in
//! callbacks to run on each signal. Use them to connect to the signals defined here.

#![allow(clippy::type_complexity)]

use std::{
    collections::{BTreeMap, btree_map},
    sync::{
        Arc,
        atomic::{AtomicU32, Ordering},
    },
};

use futures::{FutureExt, pin_mut};
use pinnacle_api_defs::pinnacle::signal::v1::{SignalRequest, StreamControl};
use tokio::sync::{
    mpsc::{UnboundedSender, unbounded_channel},
    oneshot,
};
use tokio_stream::{StreamExt, wrappers::UnboundedReceiverStream};
use tonic::Streaming;

use crate::{
    BlockOnTokio,
    input::libinput::DeviceHandle,
    output::OutputHandle,
    tag::TagHandle,
    window::{LayoutMode, WindowHandle},
};

pub(crate) trait Signal {
    type Callback;
}

macro_rules! signals {
    ( $(
        $( #[$cfg_enum:meta] )* $enum:ident => {
            $(
                $( #[$cfg:meta] )* $name:ident = {
                    enum_name = $renamed:ident,
                    callback_type = $cb:ty,
                    client_request = $req:ident,
                    on_response = $on_resp:expr,
                }
            )*
        }
    )* ) => {$(
        $(
            $( #[$cfg] )*
            pub(crate) struct $name;

            impl $crate::signal::Signal for $name {
                type Callback = $cb;
            }

            impl SignalData<$name> {
                pub(crate) fn add_callback(&mut self, callback: <$name as Signal>::Callback) -> SignalHandle {
                    if self.callback_count.load(::std::sync::atomic::Ordering::SeqCst) == 0 {
                        self.connect()
                    }

                    let Some(callback_sender) = self.callback_sender.as_ref() else {
                        unreachable!("signal should already be connected here");
                    };

                    let Some(remove_callback_sender) = self.remove_callback_sender.clone() else {
                        unreachable!("signal should already be connected here");
                    };

                    callback_sender
                        .send((self.current_id, callback))
                        .expect("failed to send callback");

                    let handle = SignalHandle::new(self.current_id, remove_callback_sender);

                    self.current_id.0 += 1;

                    handle
                }

                fn reset(&mut self) {
                    self.callback_sender.take();
                    self.dc_pinger.take();
                    self.remove_callback_sender.take();
                    self.callback_count = Default::default();
                    self.current_id = SignalConnId::default();
                }

                fn connect(&mut self) {
                    self.reset();

                    let channels = connect_signal::<_, _, <$name as Signal>::Callback, _, _>(
                        self.callback_count.clone(),
                        |out| {
                            $crate::client::Client::signal().$req(out)
                                .block_on_tokio()
                                .expect("failed to request signal connection")
                                .into_inner()
                        },
                        $on_resp,
                    );

                    self.callback_sender.replace(channels.callback_sender);
                    self.dc_pinger.replace(channels.dc_pinger);
                    self.remove_callback_sender
                        .replace(channels.remove_callback_sender);
                }
            }
        )*

        $( #[$cfg_enum] )*
        pub enum $enum {
            $( $( #[$cfg] )* $renamed($cb),)*
        }
    )*};
}

signals! {
    /// Signals relating to output events.
    OutputSignal => {
        /// An output was connected.
        ///
        /// Callbacks receive the newly connected output.
        ///
        /// FIXME: This will not run on outputs that have been previously connected.
        /// |      Tell the dev to fix this in the compositor.
        OutputConnect = {
            enum_name = Connect,
            callback_type = SingleOutputFn,
            client_request = output_connect,
            on_response = |response, callbacks| {
                let handle = OutputHandle { name: response.output_name };

                for callback in callbacks {
                    callback(&handle);
                }
            },
        }
        /// An output was connected.
        ///
        /// Callbacks receive the disconnected output.
        OutputDisconnect = {
            enum_name = Disconnect,
            callback_type = SingleOutputFn,
            client_request = output_disconnect,
            on_response = |response, callbacks| {
                let handle = OutputHandle { name: response.output_name };

                for callback in callbacks {
                    callback(&handle);
                }
            },
        }
        /// An output's logical size changed.
        ///
        /// Callbacks receive the output and new width and height.
        OutputResize = {
            enum_name = Resize,
            callback_type = Box<dyn FnMut(&OutputHandle, u32, u32) + Send + 'static>,
            client_request = output_resize,
            on_response = |response, callbacks| {
                let handle = OutputHandle { name: response.output_name };

                for callback in callbacks {
                    callback(&handle, response.logical_width, response.logical_height)
                }
            },
        }
        /// An output's location in the global space changed.
        ///
        /// Callbacks receive the output and new x and y.
        OutputMove = {
            enum_name = Move,
            callback_type = Box<dyn FnMut(&OutputHandle, i32, i32) + Send + 'static>,
            client_request = output_move,
            on_response = |response, callbacks| {
                let handle = OutputHandle { name: response.output_name };

                for callback in callbacks {
                    callback(&handle, response.x, response.y)
                }
            },
        }
        /// The pointer entered an output.
        ///
        /// Callbacks receive the output the pointer entered.
        OutputPointerEnter = {
            enum_name = PointerEnter,
            callback_type = SingleOutputFn,
            client_request = output_pointer_enter,
            on_response = |response, callbacks| {
                let handle = OutputHandle { name: response.output_name };

                for callback in callbacks {
                    callback(&handle);
                }
            },
        }
        /// The pointer left an output.
        ///
        /// Callbacks receive the output the pointer left.
        OutputPointerLeave = {
            enum_name = PointerLeave,
            callback_type = SingleOutputFn,
            client_request = output_pointer_leave,
            on_response = |response, callbacks| {
                let handle = OutputHandle { name: response.output_name };

                for callback in callbacks {
                    callback(&handle);
                }
            },
        }
        /// The window got keyboard focus.
        ///
        /// Callbacks receive the newly focused window.
        OutputFocused = {
            enum_name = Focused,
            callback_type = SingleOutputFn,
            client_request = output_focused,
            on_response = |response, callbacks| {
                let handle = OutputHandle { name: response.output_name };

                for callback in callbacks {
                    callback(&handle);
                }
            },
        }
    }
    /// Signals relating to window events.
    WindowSignal => {
        /// The pointer entered a window.
        ///
        /// Callbacks receive the window the pointer entered.
        WindowPointerEnter = {
            enum_name = PointerEnter,
            callback_type = SingleWindowFn,
            client_request = window_pointer_enter,
            on_response = |response, callbacks| {
                let handle = WindowHandle { id: response.window_id };

                for callback in callbacks {
                    callback(&handle);
                }
            },
        }
        /// The pointer left a window.
        ///
        /// Callbacks receive the window the pointer left.
        WindowPointerLeave = {
            enum_name = PointerLeave,
            callback_type = SingleWindowFn,
            client_request = window_pointer_leave,
            on_response = |response, callbacks| {
                let handle = WindowHandle { id: response.window_id };

                for callback in callbacks {
                    callback(&handle);
                }
            },
        }
        /// The window got keyboard focus.
        ///
        /// Callbacks receive the newly focused window.
        WindowFocused = {
            enum_name = Focused,
            callback_type = SingleWindowFn,
            client_request = window_focused,
            on_response = |response, callbacks| {
                let handle = WindowHandle { id: response.window_id };

                for callback in callbacks {
                    callback(&handle);
                }
            },
        }
        /// A window's title changed.
        ///
        /// Callbacks receive the window and new title.
        WindowTitleChanged = {
            enum_name = TitleChanged,
            callback_type = Box<dyn FnMut(&WindowHandle, &str) + Send + 'static>,
            client_request = window_title_changed,
            on_response = |response, callbacks| {
                let handle = WindowHandle { id: response.window_id };
                let title = response.title;

                for callback in callbacks {
                    callback(&handle, &title);
                }
            },
        }

        /// A window's layout mode changed.
        ///
        /// Callbacks receive the window and new layout mode.
        WindowLayoutModeChanged = {
            enum_name = LayoutModeChanged,
            callback_type = Box<dyn FnMut(&WindowHandle, LayoutMode) + Send + 'static>,
            client_request = window_layout_mode_changed,
            on_response = |response, callbacks| {
                let handle = WindowHandle { id: response.window_id };

                if let Ok(layout_mode) = response.layout_mode().try_into() {
                    for callback in callbacks {
                        callback(&handle, layout_mode);
                    }
                }
            },
        }


        /// A window was created (i.e., mapped for the first time).
        ///
        /// Callbacks receive the newly created window.
        WindowCreated = {
            enum_name = Created,
            callback_type = SingleWindowFn,
            client_request = window_created,
            on_response = |response, callbacks| {
                let handle = WindowHandle { id: response.window_id };
                for callback in callbacks {
                    callback(&handle);
                }
            },
        }

        /// A window was closed.
        ///
        /// Callbacks receive the window that was just closed, its title, and its app_id.
        /// Note: The window handle is no longer valid as the window was destroyed.
        /// Any subsequent operations on this handle will likely fail.
        WindowDestroyed = {
            enum_name = Destroyed,
            callback_type = Box<dyn FnMut(&WindowHandle, &str, &str) + Send + 'static>,
            client_request = window_destroyed,
            on_response = |response, callbacks| {
                let handle = WindowHandle { id: response.window_id };
                let title = response.title;
                let app_id = response.app_id;

                for callback in callbacks {
                    callback(&handle, &title, &app_id);
                }
            },
        }
    }
    /// Signals relating to tag events.
    TagSignal => {
        /// A tag was set to active or not active.
        TagActive = {
            enum_name = Active,
            callback_type = Box<dyn FnMut(&TagHandle, bool) + Send + 'static>,
            client_request = tag_active,
            on_response = |response, callbacks| {
                let handle = TagHandle { id: response.tag_id };

                for callback in callbacks {
                    callback(&handle, response.active);
                }
            },
        }
        /// A tag was created.
        TagCreated = {
            enum_name = Created,
            callback_type = Box<dyn FnMut(&TagHandle) + Send + 'static>,
            client_request = tag_created,
            on_response = |response, callbacks| {
                let handle = TagHandle { id: response.tag_id };

                for callback in callbacks {
                    callback(&handle);
                }
            },
        }
        /// A tag was removed
        TagRemoved = {
            enum_name = Removed,
            callback_type = Box<dyn FnMut(&TagHandle) + Send + 'static>,
            client_request = tag_removed,
            on_response = |response, callbacks| {
                let handle = TagHandle { id: response.tag_id };

                for callback in callbacks {
                    callback(&handle);
                }
            },
        }
    }
    /// Signals relating to input events.
    InputSignal => {
        /// A new input device was connected.
        InputDeviceAdded = {
            enum_name = DeviceAdded,
            callback_type = Box<dyn FnMut(&DeviceHandle) + Send + 'static>,
            client_request = input_device_added,
            on_response = |response, callbacks| {
                let handle = DeviceHandle { sysname: response.device_sysname };

                for callback in callbacks {
                    callback(&handle);
                }
            },
        }
    }
}

pub(crate) type SingleOutputFn = Box<dyn FnMut(&OutputHandle) + Send + 'static>;
pub(crate) type SingleWindowFn = Box<dyn FnMut(&WindowHandle) + Send + 'static>;

pub(crate) struct SignalState {
    pub(crate) output_connect: SignalData<OutputConnect>,
    pub(crate) output_disconnect: SignalData<OutputDisconnect>,
    pub(crate) output_resize: SignalData<OutputResize>,
    pub(crate) output_move: SignalData<OutputMove>,
    pub(crate) output_pointer_enter: SignalData<OutputPointerEnter>,
    pub(crate) output_pointer_leave: SignalData<OutputPointerLeave>,
    pub(crate) output_focused: SignalData<OutputFocused>,

    pub(crate) window_pointer_enter: SignalData<WindowPointerEnter>,
    pub(crate) window_pointer_leave: SignalData<WindowPointerLeave>,
    pub(crate) window_focused: SignalData<WindowFocused>,
    pub(crate) window_title_changed: SignalData<WindowTitleChanged>,
    pub(crate) window_layout_mode_changed: SignalData<WindowLayoutModeChanged>,
    pub(crate) window_created: SignalData<WindowCreated>,
    pub(crate) window_destroyed: SignalData<WindowDestroyed>,

    pub(crate) tag_active: SignalData<TagActive>,
    pub(crate) tag_created: SignalData<TagCreated>,
    pub(crate) tag_removed: SignalData<TagRemoved>,

    pub(crate) input_device_added: SignalData<InputDeviceAdded>,
}

impl std::fmt::Debug for SignalState {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("SignalState").finish()
    }
}

impl SignalState {
    pub(crate) fn new() -> Self {
        Self {
            output_connect: SignalData::new(),
            output_disconnect: SignalData::new(),
            output_resize: SignalData::new(),
            output_move: SignalData::new(),
            output_pointer_enter: SignalData::new(),
            output_pointer_leave: SignalData::new(),
            output_focused: SignalData::new(),

            window_pointer_enter: SignalData::new(),
            window_pointer_leave: SignalData::new(),
            window_focused: SignalData::new(),
            window_title_changed: SignalData::new(),
            window_layout_mode_changed: SignalData::new(),
            window_created: SignalData::new(),
            window_destroyed: SignalData::new(),

            tag_active: SignalData::new(),
            tag_created: SignalData::new(),
            tag_removed: SignalData::new(),

            input_device_added: SignalData::new(),
        }
    }

    pub(crate) fn shutdown(&mut self) {
        self.output_connect.reset();
        self.output_disconnect.reset();
        self.output_resize.reset();
        self.output_move.reset();
        self.output_pointer_enter.reset();
        self.output_pointer_leave.reset();
        self.output_focused.reset();

        self.window_pointer_enter.reset();
        self.window_pointer_leave.reset();
        self.window_focused.reset();
        self.window_title_changed.reset();
        self.window_layout_mode_changed.reset();
        self.window_created.reset();
        self.window_destroyed.reset();

        self.tag_active.reset();
        self.tag_created.reset();
        self.tag_removed.reset();

        self.input_device_added.reset();
    }
}

#[derive(Default, Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) struct SignalConnId(pub(crate) u32);

pub(crate) struct SignalData<S: Signal> {
    callback_sender: Option<UnboundedSender<(SignalConnId, S::Callback)>>,
    remove_callback_sender: Option<UnboundedSender<SignalConnId>>,
    dc_pinger: Option<oneshot::Sender<()>>,
    current_id: SignalConnId,
    callback_count: Arc<AtomicU32>,
}

impl<S: Signal> SignalData<S> {
    fn new() -> Self {
        Self {
            callback_sender: Default::default(),
            remove_callback_sender: Default::default(),
            dc_pinger: Default::default(),
            current_id: Default::default(),
            callback_count: Default::default(),
        }
    }
}

struct ConnectSignalChannels<F> {
    callback_sender: UnboundedSender<(SignalConnId, F)>,
    dc_pinger: oneshot::Sender<()>,
    remove_callback_sender: UnboundedSender<SignalConnId>,
}

fn connect_signal<Req, Resp, F, T, O>(
    callback_count: Arc<AtomicU32>,
    to_in_stream: T,
    mut on_response: O,
) -> ConnectSignalChannels<F>
where
    Req: SignalRequest + Send + 'static,
    Resp: Send + 'static,
    F: Send + 'static,
    T: FnOnce(UnboundedReceiverStream<Req>) -> Streaming<Resp>,
    O: FnMut(Resp, btree_map::ValuesMut<'_, SignalConnId, F>) + Send + 'static,
{
    let (control_sender, recv) = unbounded_channel::<Req>();
    let out_stream = UnboundedReceiverStream::new(recv);

    let mut in_stream = to_in_stream(out_stream);

    let (callback_sender, mut callback_recv) = unbounded_channel::<(SignalConnId, F)>();
    let (remove_callback_sender, mut remove_callback_recv) = unbounded_channel::<SignalConnId>();
    let (dc_pinger, mut dc_ping_recv) = oneshot::channel::<()>();

    let signal_future = async move {
        let mut callbacks = BTreeMap::<SignalConnId, F>::new();

        control_sender
            .send(Req::from_control(StreamControl::Ready))
            .map_err(|err| {
                println!("{err}");
                err
            })
            .expect("send failed");

        loop {
            let in_stream_next = in_stream.next().fuse();
            pin_mut!(in_stream_next);
            let callback_recv_recv = callback_recv.recv().fuse();
            pin_mut!(callback_recv_recv);
            let remove_callback_recv_recv = remove_callback_recv.recv().fuse();
            pin_mut!(remove_callback_recv_recv);
            let mut dc_ping_recv_fuse = (&mut dc_ping_recv).fuse();

            futures::select! {
                response = in_stream_next => {
                    let Some(response) = response else { continue };

                    match response {
                        Ok(response) => {
                            on_response(response, callbacks.values_mut());

                            control_sender
                                .send(Req::from_control(StreamControl::Ready))
                                .expect("send failed");

                            tokio::task::yield_now().await;
                        }
                        Err(status) => eprintln!("Error in recv: {status}"),
                    }
                }
                callback = callback_recv_recv => {
                    if let Some((id, callback)) = callback {
                        callbacks.insert(id, callback);
                        callback_count.fetch_add(1, Ordering::SeqCst);
                    }
                }
                remove = remove_callback_recv_recv => {
                    if let Some(id) = remove {
                        if callbacks.remove(&id).is_some() {
                            assert!(callback_count.fetch_sub(1, Ordering::SeqCst) > 0);
                        }
                        if callbacks.is_empty() {
                            assert!(callback_count.load(Ordering::SeqCst) == 0);
                            control_sender.send(Req::from_control(StreamControl::Disconnect)).expect("send failed");
                            break;
                        }
                    }
                }
                _dc = dc_ping_recv_fuse => {
                    let _ = control_sender.send(Req::from_control(StreamControl::Disconnect));
                    break;
                }
            }
        }
    };

    tokio::spawn(signal_future);

    ConnectSignalChannels {
        callback_sender,
        dc_pinger,
        remove_callback_sender,
    }
}

/// A handle that can be used to disconnect from a signal connection.
///
/// This will remove the connected callback.
#[derive(Debug, Clone)]
pub struct SignalHandle {
    id: SignalConnId,
    remove_callback_sender: UnboundedSender<SignalConnId>,
}

impl SignalHandle {
    pub(crate) fn new(
        id: SignalConnId,
        remove_callback_sender: UnboundedSender<SignalConnId>,
    ) -> Self {
        Self {
            id,
            remove_callback_sender,
        }
    }

    /// Disconnects this signal connection.
    pub fn disconnect(&self) {
        self.remove_callback_sender
            .send(self.id)
            .expect("failed to disconnect signal");
    }
}
